KEP-4631: LoadBalancer Service Status Improvements, initial proposal

danwinship opened the pull request:
<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Initial proposal for KEP-4631

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4631

<!-- other comments or additional information -->
- Other comments:

```
While updating the e2e load balancer tests after the final removal of
in-tree cloud providers, we have run into three problems:

  1. The tests have hard-coded timeouts (that sometimes differ per
     cloud provider) for deciding how long to wait for the cloud
     provider to update the service. It would make much more sense for
     the cloud provider to just provide information about its status
     on the Service object, so the tests could just monitor that.

  2. The tests recognize that not all cloud providers can implement
     all load balancer features, but in the past this was handled by
     hard-coding the information into the individual tests. (e.g.,
     `e2eskipper.SkipUnlessProviderIs("gce", "gke", "aws")`) These
     skip rules no longer work in the providerless tree, and this
     approach doesn't scale anyway. OTOH, we don't want to have to
     provide a separate `Feature:` tag for each load balancer
     subfeature, or have each cloud provider have to maintain their
     own set of `-ginkgo.skip` rules. It would be better if the e2e
     tests themselves could just figure out, somehow, whether they
     were running under a cloud provider that intends to implement the
     feature they are testing, or a cloud provider that doesn't.

  3. In some cases, because the existing tests were only run on
     certain clouds, it is not clear what the expected semantics are
     on other clouds. For example, since `IPMode: Proxy` load
     balancers can't preserve the client source IP in the way that
     `ExternalTrafficPolicy: Local` expects, should they refuse to
     provision a load balancer at all, or should they provision a load
     balancer that fails to preserve the source IP?

This KEP proposes new additions to `service.Status.LoadBalancer` and
`service.Status.Conditions` to allow cloud providers to better
communicate the status of load balancer support and provisioning, and
new guidelines on how cloud providers should handle load balancers for
services that they cannot fully support.
```

/assign @aojea @thockin

danwinship commented:
/sig cloud-provider
/sig testing

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+     approach doesn't scale anyway. OTOH, we don't want to have to
+     provide a separate `Feature:` tag for each load balancer
+     subfeature, or have each cloud provider have to maintain their
+     own set of `-ginkgo.skip` rules. It would be better if the e2e
```
I don't like "smart tests" that have different execution flows depending on what... current e2e for loadbalancers are not good tests as try to test a lot of different things in one execution ... the test should assert on a feature and a behavior, we may need to break existing tests down to see if we still need to do this

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+
+  3. In some cases, because the existing tests were only run on
+     certain clouds, it is not clear what the expected semantics are
+     on other clouds. For example, since `IPMode: Proxy` load
```
@thockin this was an interesting finding during the implementation of this mode in cloud-provider-kind, we need to try to flesh out more details before going to GA

danwinship replied:
The problem isn't with the `IPMode` KEP; the distinction between "proxy mode" and "VIP mode" already existed before that; it's just that before `IPMode` made it explicit, it was controlled implicitly by whether the LB set `Hostname` or `IP`.

(Which is to say, even if we dropped `IPMode`, the problem would still exist. For example, with the default cloud-provider-aws load balancers.)

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+
+### Goals
+
+- Allow cloud providers to indicate that they are working on
+  provisioning load balancer infrastructure, so that
+  users/operators/tests can distinguish the case of "it is taking a
+  while for the cloud to provision the load balancer" from "the cloud
+  has failed to provision a load balancer" and "there is no cloud
+  provider so load balancers don't work".
```
❤️

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+  provision a particular `LoadBalancer` service, and why.
+
+- Allow cloud providers to indicate when they have provided an
+  "imperfect" load balancer that the user may or may not consider to
```
kind of degraded mode?

cc: @bowei

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+
+- Allow users and e2e tests to either prevent or reliably recognize
+  the case where the cloud provider has provisioned a load balancer
+  without support for a particular feature the user/test needs (even
```
we did this for the TCP and UDP loadbalancers once we relaxed the validation on the apiserver https://github.com/kubernetes/kubernetes/pull/115966 , it was really painful to implement this

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+  other providers to follow.
+
+- Update the e2e load balancer tests to use the new functionality to
+  determine which tests to skip, so that any cloud provider that
```
I agree on the spirit disagree on the goal, if we want to define a behavior that given Status X it should be a noop we do a test that do exactly that, but we should not do a test that starts asserting Z and given Status X or Y behave different

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+it make sense for them to automatically switch to "NLB" load balancers
+if the Service is `ExternalTrafficPolicy: Local`?
+
+If so, should we add a field to `LoadBalancerStatus` (or
+`LoadBalancerIngress`) to indicate the LB type that was chosen?
```
unfortunately the current state of Types is based on annotations, so this will not be feasible, as users create the service with annotations foo and bar, so it explicitly indicates the intention of using the loadbalancer of that kind ... if the controller goes its way and change these annotations it can also create a hotloop situation

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+required that "All of the major clouds support \[mixed protocols] or
+indicate non-support properly" as a condition for graduating to
+Beta, it seems that only `cloud-provider-gcp` was ever updated to set
+this condition, and that was several months after the feature went
```
https://github.com/kubernetes/enhancements/pull/4632/files#r1598418132 and because it was my personal choice, otherwise it will not be implemented ... we should assume not much help here from implementations

danwinship replied:
> we should assume not much help here from implementations

well, the idea is that if they want to run the `[Feature:LoadBalancer]` e2e tests, they'd need to implement whatever we do

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+so this feels "un-Kubernetes-like", though at the same time, it's not
+like the current Kubernetes networking configuration situation is
+really great, and there has been some discussion of trying to provide
+more explicit and well-defined cluster networking configuration in the
```
yeah, this is in the "train has left the station" section :)

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
````diff
@@ -0,0 +1,1210 @@
+<<[/UNRESOLVED]>>
+```
+
+#### The `LoadBalancerServing` Condition
````
Can we arrange it so that, as much as possible, the behavior and names are the same between Services and Gateways?

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+update, even if the value of `LoadBalancerProvisioning` remains
+`False`.
+
+#### Terminating Condition
```
I'm not conviced this is useful - what does a user do with this information?

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+     indicating that the load balancer for the service is already
+     available.
+
+  3. The cloud provider sets `LoadBalancerProvisioning=True`,
```
What about `(true, true)` ?  Conditions are best when they are orthogonal.  These are not.  We have to spec it, since both fields COULD be set at the same time.

IOW, don't let this become a state-machine API.  It's a collection of roughly independent observations.

I see it more clearly below, so now this feels duplicative

danwinship replied:
This is actually specified below; (true, true) means it is both serving and provisioning. ie, it is continuing to serve while reprovisioning for an update.

> IOW, don't let this become a state-machine API. It's a collection of roughly independent observations.

I feel like we definitely want the "provisioning" observation. I feel like the "serving" observation is also useful? Did you have some other idea for conditions?

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+If the `LoadBalancerServing` and `LoadBalancerProvisioning` conditions
+do not appear on the Service within a reasonable amount of time after
+the Service is created (or modified to be `Type: LoadBalancer`), the
+user should assume that no cloud provider is running, and that
```
Or that the cloud provider does not implement this extension.  I don't think we can say more than that.

danwinship commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+  - The Service has `AllocateLoadBalancerNodePorts: false`, but the
+    cloud only supports NodePort-based load balancing.
+
+  - The Service is `ExternalTrafficPolicy: Local` but the cloud cannot
```
> eTP=Local means what it says

lol, I thought that in the whole "what things should be traffic policy vs what things should be topology" debate (around PreferLocal) we had agreed that eTP=Local _doesn't_ mean what it says, because the actual intent of the feature is "preserve client source IP", not "route traffic in a particular way"; the routing is purely a side effect of making it possible to implement "preserve client source IP". (IOW an implementation of eTP:Local that preserves client IP while doing routing in an unexpected way would be compliant, but an implementation that routes "correctly" but loses client IP is not.)

> I think this is not really about IP preservation but has everything to do with not implementing `hCNP` or some equivalent mechanism.

What I was saying in that example was, if client IP preservation is considered a mandatory-to-implement aspect of eTP:Local, then there's no point in proxy-ish LBs implementing eTP:Local at all, and thus no reason for them to implement HCNP. (Whereas for VIP-ish LBs, there is really no good argument for not implementing HCNP.)

(If we don't think that client IP preservation is required for eTP:Local, then we should just assume all LBs will implement HCNP, and they're just buggy if they don't.)

danwinship commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+    supports single-stack load balancers, so it would only be able to
+    serve clients of one IP family.
+
+  - The Service is `ExternalTrafficPolicy: Local` but the cloud cannot
```
I think actually no. Azure has this trick where the entire cloud network is aware of the load balancer NAT state, so the LB can DNAT packets to a NodePort without masquerading them, and then the reply packet will get un-DNAT-ed correctly even though logically speaking it doesn't pass through the LB.

danwinship commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+    provide "efficient"/"one-hop" load balancing *without* source IP
+    preservation (which might actually be all the user wanted anyway).
+
+  - The Service uses `LoadBalancerSourceRanges` but the cloud does not
```
Yes, but kube-proxy's implementation will only work if the LB preserves source IP.

danwinship commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1210 @@
+
+[Kubernetes Network Configuration]: https://docs.google.com/document/d/1Dx7Qu5rHGaqoWue-JmlwYO9g_kgOaQzwaeggUsLooKo/edit
+
+#### Option 3: Make the Cloud Provider Indicate Dynamically Which Features It Has Implemented
```
We've definitely talked about this sort of thing with NetworkPolicy.

But it's true... the service proxy could indicate which features it has implemented too.

danwinship commented on keps/sig-network/4631-loadbalancerstatus/README.md:
````diff
@@ -0,0 +1,1210 @@
+request that explicitly:
+
+```
+# The load balancer must be able to support mixed TCP and UDP, and
````
only in the case where they don't know for sure what the cloud supports, and they want to require failure if it doesn't support what they want.

danwinship commented:
Updates:
- Updated for review comments, started filling in the rest of the template ("Test Plan", "Graduation", "Version Skew Strategy", PRR, etc)
- Removed `UNRESOLVED auto-skipping`: Antonio agrees it's reasonable to have "tri-state" tests (pass/fail/skip) as long as the semantics are explicit. In some cases this will require rewriting the tests to put the "objectionable" bits at the start, so we can always skip immediately after the initial LB provisioning if the cloud doesn't support the feature.
- The version skew section made me think about the behavior when a cloud provider is updated and finds itself in a cluster with pre-existing "bad" load balancers, and whether we should maybe add an explicit "unsupported" or "broken" condition.
- Added a summary section to the top of the "Expected Behavior When the Cloud Provider Doesn't Know That It Can't Implement a Load Balancer" section, which I think is the big open question before this becomes `implementable`.
- Added detailed notes to the e2e Test Plan section clarifying what skips will be needed for which tests. Also, while writing out the Test Plan section, I realized that to avoid regressing coverage, we're basically going to have to fork the LB tests, so we can have one hacky GCE-and-kind-only set, and one KEP-4631 set, which will initially be Alpha, but which will eventually replace the hacky ones.

soltysh commented:
Based on [this comment](https://github.com/kubernetes/enhancements/issues/4631#issuecomment-2318318699) from Tim, I'm assuming the work will happen for 1.32. Please don't forget to fill in the PRR, and its file (see https://github.com/kubernetes/enhancements/blob/master/keps/prod-readiness/template/nnnn.yaml) for this KEP. Feel free to assign me for that approval.

danwinship commented:
OK, big update, mostly resolving the "when the cloud provider can't implement a feature" question. I went with `service.spec.requiredLoadBalancerFeatures`, which I think works best overall.

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+
+- Helpers for the new API implemented in `k8s.io/cloud-provider`
+
+- cloud-provider-kind and cloud-provider-gcp set conditions
```
I don't think there is people today to work with cloud-provider-gcp and review the changes, so this is not likely going to happen and cloud-provider-kind can not be the only thing implementing it

danwinship replied:
oh... hm, that's sort of annoying if true since it means we can't really test this fully within k8s CI. We can change this to say AWS or Azure instead, but who will actually end up testing it in that case?

aojea replied:
this is the problem with Service loadbalancer, the last changes added to status, like IPMode and the relax validation on multiple TCP and UDP ports (it seems only Azure) were never implemented and they are not likely going to be implemented so I don't know if this is worth the effort ... in theory Gateway API is the things should be replacing this Service type

soltysh commented:
Ad reminder that the PRR file is required for this PR
/hold
to make sure we don't merge that w/o it

danwinship commented:
Yeah, I haven't done the PRR part of the KEP yet because we hadn't actually decided on the scope of the work yet. (The enhancements freeze snuck up on my and I'm assuming I've missed it for 1.32.)

soltysh commented:
> Yeah, I haven't done the PRR part of the KEP yet because we hadn't actually decided on the scope of the work yet. (The enhancements freeze snuck up on my and I'm assuming I've missed it for 1.32.)

There's still time, today and tomorrow. The [freeze is 19:00 PDT Thursday 10th October 2024](https://github.com/kubernetes/sig-release/tree/master/releases/release-1.32#summary) :wink:

soltysh commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+
+###### Are there any tests for feature enablement/disablement?
+
+TODO (we should make sure the cloud provider updates pre-existing
```
This answer is required for alpha.

soltysh commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+Pick one more of these and delete the rest.
+-->
+
+- [ ] Metrics
```
Not required for alpha, but it's nice to think through those.

soltysh commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+It will add 2 or 3 new conditions to Services of `Type: LoadBalancer`, and
+one new optional spec field.
```
Nit: can you estimate the size?

soltysh commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
```
This should be filled in.

soltysh commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
```
Would be nice to have this.

soltysh reviewed (commented):
Mostly the alpha PRR looks good, there are a few minor bits in tests (units + enablement/disablement) and that alpha criteria comment from Antonio.

danwinship commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+    provisioned, because of a problem with the underlying cloud
+    infrastructure (explained further in `Message`).
+
+#### The `LoadBalancerProvisioning` Condition
```
The difference between the two is that Provisioning tells you "is the cloud provider currently trying to bring the cloud infrastructure in line with the Service definition?" and Serving tells you "is there a working cloud load balancer?"

Note that [all four combinations of values are meaningful](https://github.com/kubernetes/enhancements/pull/4632/files#diff-fb7b4f95f33aa05e148de4dee51717304206ff0dfea2d81908549baa62be9f97R359-R374).

Also, having a separate Provisioning condition lets you `kubectl wait --for condition=LoadBalancerProvisioning=False`, which you couldn't do if the provisioning was a "step" of the Serving condition. (Well, you'd have to use `--for jsonpath='...'` with some expression to find a LoadBalancerServing condition (of any Value) whose Reason isn't Provisioning.)

All that said, Tim had the same objection so either I'm explaining this badly, or else no one else thinks the distinction is useful...

soltysh replied:
I've seen that text, and now it makes more sense, thank you. What confused me was that in [this section](https://github.com/kubernetes/enhancements/pull/4632/files#diff-fb7b4f95f33aa05e148de4dee51717304206ff0dfea2d81908549baa62be9f97R359-R374) you've used `Provisioning` and `Serving` which I assumed are the reason, not the whole condition. What would make it explicitly clear is to use the full name of the condition in that text.

thockin replied:
I am fine with them, after reading that response, but I still want clarify on Serving (see above)

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+balancer is handling connections, and `False` if it is not. (There are
+no defined semantics for an `Unknown` value of this condition.)
+
+(This means that at construction time, the `LoadBalancerServing`
```
Pedantically, the `.Status.LoadBalancer` field means "I have an IP allocated".  It doesn't mean that traffic is flowing or that the IP is reachable yet.

For example, it may take seconds or even minutes for infrastructure to actually start receiving and routing traffic.

Does "LoadBalancerServing" mean that we have provisioned the LB or that someone has confirmed that the traffic is flowing (from where?  what about LB source ranges?)  

Does this line up with Gateway's conditions?

danwinship replied:
Somewhere else I claim that the e2e tests will expect to be able to connect immediately upon `LoadBalancerServing` becoming `True` (though we could change that).

I think my intuition was "Serving means that the cloud provider _believes_ that the LB is accepting/redirecting connections". It definitely doesn't imply that the cloud provider has actually tried making a connection. (In addition to source ranges, there's the problem that maybe the cluster end of the service is currently not-ready, which shouldn't affect the LB condition.)

If there is internal status information that tells the cloud provider that the LB has been allocated but is not yet actually serving, then the cloud provider should not mark the service as LoadBalancerServing until the internal status information says it's serving.

If there's an allocated-but-not-yet-serving state, but the cloud provider _doesn't_ have any way of figuring that out other than by trying to connect to the LB, then I guess it needs to call that state "serving"...

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+way, the cloud provider must update at least the `LastTransitionTime`
+of the `LoadBalancerServing` condition. (For example, if the user
+changes the service's `LoadBalancerSourceRanges`, and the cloud
+provider updates the load balancer for this, it must update the
```
This is not generally a requirement of Conditions - I get why that might be valuable information but it seems wrong to overload "LastTransitionTime" to mean anything but "the last time this changed state"

danwinship replied:
ok

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+object, and `False` if it not. (There are no defined semantics for an
+`Unknown` value of this condition.)
+
+Any time any load-balancer-related field (or provider-specific
```
other fields are not really timestamped, so unless you know what time you made a change, this semantic is meaningless.  If you neeeeed to know this, we should set something atomically in the REST handler and force providers to clear it (same as we are doing for in-place pod updates)

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+
+When a load balancer Service is first created (or a non-load-balancer
+Service is converted to `Type: LoadBalancer`), one of three things
+should immediately happen:
```
how immediately?  If it takes 10 seconds for the controller to act, is that OK?

danwinship replied:
Why would the controller take 10 seconds to act?

The intent is "the controller should just do this before it starts doing anything potentially-time-consuming".

In the e2e tests we currently assume that if you make a change to a pod or service, that the endpointslice controller will act on that change "immediately" and the test will be able to observe a change to the corresponding EndpointSlice object within 30 seconds. I was planning to use the same requirement for cloud-providers setting the intitial conditions.

aojea replied:
The loadbalancer controller interacts with the cloud API and infra, this calls to the cloud APIs use to be sync and block the workers ... on very large clusters this can take ~ 1 hour per LoadBalancer.
In our scale tests for loadbalancers we have to define at least the same number of workers in the loadbalancer controller as the number of loadbalancers that are going to be created in parallel on the tests, otherwise, you are not able to process any event until some worker is available.

danwinship replied:
This KEP would (necessarily) make that architecture "illegal". One of the major points of the conditions is to be able to distinguish between "no cloud provider is running" and "the cloud provider is taking an hour to allocate your load balancer" (and "the cloud provider already determined that it's not going to be able to allocate your load balancer for some reason"). So the cloud provider needs an informer event loop that doesn't block on the cloud API.

aojea replied:
there is no need to change any architecture at all, is just the conditions need to be updated before it gets to the worker ...

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+because not all cloud providers will support the new conditions.
+However, since supporting these conditions will be a prereq for
+running the `[Feature:LoadBalancer]` e2e tests, we assume that
+eventually all cloud providers will support them.)
```
This sounds suspiciously like a breaking change

danwinship replied:
none of the tests are conformance and we don't have any stability guarantees around e2e tests.

But that was written when I had hoped to get this into 1.31, and given that that obviously didn't happen, it's probably better to leave the `[Feature:LoadBalancer]` tag with its current meaning ("run LB tests under the assumption that the cloud is cloud-provider-gcp; you can manually skip things if you aren't"), and add a new tag for the new ones.

thockin reviewed (commented):
I note that this does not have a PRR file.

Sending first-half notes now, will finish reading in a bit

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+running the `[Feature:LoadBalancer]` e2e tests, we assume that
+eventually all cloud providers will support them.)
+
+When the user updates load-balancer-relevant fields in a Service that
```
I assume LB impls should also remove these if the service is updated in a way that removes the LB?

thockin commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+    requested session affinity but the cloud provider has provisioned
+    a load balancer that does not support it.
+
+  - `Status=True, Reason=ExternalTrafficPolicyNotSupported`: The
```
I'm not sure on this one.  IPMode=Proxy already covers this, no?

danwinship replied:
IPMode=VIP is necessary for implementing ExternalTrafficPolicy=Local but strictly speaking it's not sufficient. (The source IP could be lost for some other reason.)

Also, `IPMode` was intended as cloud-provider-to-service-proxy communication, while the conditions are cloud-provider-to-end-user communication. So even if they're technically redundant, they exist for different reasons?

I agree that it's not clearly necessary though. I could drop it.

thockin reviewed (commented):
I let this slip, and now it seems too late, but I am OK with the vast majority of this proposal.

/approve

aojea commented:
> I let this slip, and now it seems too late, but I am OK with the vast majority of this proposal.
> 
> /approve

/hold

The higher risk here is adding fields that nobody implements, @danwinship and I talked offline, and Dan has a good  idea to simplify this proposal that can achieve the goals and solve the issues detailed here

aojea commented on keps/sig-network/4631-loadbalancerstatus/README.md:
```diff
@@ -0,0 +1,1066 @@
+
+We will add three new Service conditions:
+
+  - `LoadBalancerServing`, to indicate explicitly that the load
+    balancer is (or isn't) serving traffic.
```
I kind of prefer `LoadBalancerReady ` here, no strong opinion

aojea commented:
agree, with the reduced scope it can be also good for a new contributor that want to drive it through the KEP process ...

soltysh commented on keps/sig-network/4631-loadbalancerstatus/kep.yaml:
```diff
@@ -0,0 +1,45 @@
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.32"
```
Update the versions

soltysh reviewed (commented):
Comments from last PRR review still hold, don't forget the PRR `4631.yaml` file.

thockin reviewed (commented):
I found that I had unsubmitted comments :(
